### PR TITLE
[SPARK-7768][SQL] Revise user defined types (UDT)

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

This patch is going to revise the current API for User Defined Types (UDTs).

Currently the API of `UserDefinedType` asks developers to implement the method that fills user data into SparkSQL's internal format such as `GenericInternalRow`, `UnsafeArrayData`, etc.

One target of this patch is to simplify the way to write UDTs. Developers can just use Scala types, instead of internal types such as ArrayData, when writing UDTs. `UserDefinedType` will use SparkSQL's encoder to convert user data to internal format.

For example, the following is the serialization method of `MatrixUDT`.

Before this patch:

    override def serialize(obj: Matrix): InternalRow = {
      val row = new GenericInternalRow(7)
      obj match {
        case sm: SparseMatrix =>
          row.setByte(0, 0)
          row.setInt(1, sm.numRows)
          row.setInt(2, sm.numCols)
          row.update(3, UnsafeArrayData.fromPrimitiveArray(sm.colPtrs))
          row.update(4, UnsafeArrayData.fromPrimitiveArray(sm.rowIndices))
          row.update(5, UnsafeArrayData.fromPrimitiveArray(sm.values))
          row.setBoolean(6, sm.isTransposed)

        case dm: DenseMatrix =>
          row.setByte(0, 1)
          row.setInt(1, dm.numRows)
          row.setInt(2, dm.numCols)
          row.setNullAt(3)
          row.setNullAt(4)
          row.update(5, UnsafeArrayData.fromPrimitiveArray(dm.values))
          row.setBoolean(6, dm.isTransposed)
      }
      row
    }

After this patch:

    override def writeRow(obj: Matrix): Row = {
      obj match {
        case sm: SparseMatrix =>
          Row(0.toByte, sm.numRows, sm.numCols, sm.colPtrs, sm.rowIndices, sm.values, sm.isTransposed)

        case dm: DenseMatrix =>
          Row(1.toByte, dm.numRows, dm.numCols, null, null, dm.values, dm.isTransposed)
      }
    }

Developers now manipulate external row `Row` and Scala types. Encoder will take care of converting the data to SparkSQL's internal format.

### Main API change

In the past, two main methods are needed to be implemented in developers' UDTs which extend `UserDefinedType`.

    /** Convert the user type to a SQL datum */
    def serialize(obj: UserType): Any

    /** Convert a SQL datum to the user type */
    def deserialize(datum: Any): UserType

Developers put/get the data of user class into/from an internal type of Spark SQL.

Now developers after this patch, are changed to implement two methods which put/get user data into/from an external row of SparkSQL.

    /** Convert the object of user type to an external row. Must be implemented in subclasses. */
    def writeRow(obj: UserType): Row

    /** Convert the external row to an object of user type. Must be implemented in subclasses. */
    def readRow(row: Row): UserType

### Benchmark

Ran a benchmark against previous `UserDefinedType`.

Code:

    private val random = new Random(100)
    private lazy val pointsRDD = (0 to 1000).map { i =>
      val features = (0 to 100).map { _ =>
        random.nextDouble()
      }
      MyLabeledPoint(i % 10, new UDT.MyDenseVector(features.toArray))
    }.toDF()

    test("serialize MyLabeledPoint") {
      val N = 10L << 5
      runBenchmark("serialize MyLabeledPoint", N) {
        pointsRDD.groupBy('label).agg(max('features)).collect()
      }
    }

Before this patch:

    Java HotSpot(TM) 64-Bit Server VM 1.8.0_102-b14 on Linux 4.4.39-moby
    Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
    serialize MyLabeledPoint:                Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    serialize MyLabeledPoint wholestage off        271 /  328          0.0      845767.7       1.0X
    serialize MyLabeledPoint wholestage on         131 /  261          0.0      408173.6       2.1X

After this patch:

    Java HotSpot(TM) 64-Bit Server VM 1.8.0_102-b14 on Linux 4.4.39-moby
    Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
    serialize MyLabeledPoint:                Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    serialize MyLabeledPoint wholestage off        221 /  331          0.0      692100.4       1.0X
    serialize MyLabeledPoint wholestage on         125 /  145          0.0      391120.1       1.8X

Basically, after this patch the serialization/deserialization of `UserDefinedType` can compete or be slightly better than previous one.

## How was this patch tested?

Existing Jenkins tests.

### CODE REVIEW

**Comment by shaneknapp** [comment]
retest this please

**Comment by viirya** [comment]
cc @marmbrus @rxin @mengxr User defined types (UDT) is a critical feature to developers. As we have marked current UDT API private for a while, developers can't use UDT and that is a problem for them. I would like to try to move it forward.

So may you please take a look this and provide your suggestion whether this is a good direction for you?

Thank you.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UDTRegistration.scala [hunk 1]
```diff
@@ -29,7 +29,6 @@ import org.apache.spark.util.Utils
  * However, by doing this, we add SparkSQL dependency on user classes. This object provides
  * alternative approach to register UDTs for user classes.
  */
-private[spark]
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UDTRegistration.scala]: Alternatively, we can add an interface to `UDTRegistration`, instead of making it public.

**Comment by viirya** [comment]
ping @rxin @marmbrus  Can you take a brief look and let me know what you think about this change to UDT? Thank you.

**Comment by viirya** [comment]
also cc @cloud-fan @hvanhovell

**Comment by MLnick** [comment]
Is there no SQL committer support for this? Seems like a critical feature for Spark users with no response from any SQL folks. 

Making UDT public in some way is pretty important no?

**Comment by rxin** [comment]
I don't know how important it is. It seems like it's primarily used by MLlib and very few other things ...

**Comment by MLnick** [comment]
The JIRA has 37 watchers (ranked 17 across all Spark JIRAs) and 22 votes (ranked 8 overall across all JIRAs). I'd have thought that's some indication of user interest.

**Comment by lossyrob** [comment]
Using geospatial data types with Spark SQL require UDTs, and solutions are hacking around the fact that they are private to the package. e.g.:

- https://github.com/s22s/geotrellis-spark-sql/blob/master/src/main/scala/org/apache/spark/sql/gt/types/AbstractTileUDT.scala#L32
- https://github.com/locationtech/geomesa/blob/master/geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLTypes.scala#L70

[GeoTrellis](https://geotrellis.io/) and [GeoMesa](http://www.geomesa.org/) wrote this code as a way to get around UDTs not being public available, with the idea that we'd fix things up once the API went back to being public. As you can see by how much code is written against it, it is very useful and it works. We'd really like to see this feature and API become public as it's very essential to our work.

Let me know if there's any help I can give.

/cc @jnh5y @metasim

**Comment by metasim** [comment]
Thanks @lossyrob for sharing those examples here. As a followup, I want to point out that my experience extending `UserDefinedType[_]`, as an external user, has been very positive, and feels like a very mature and well designed API, as it stands. I hope that it is opened up soon, at least as an `@Experimental` API.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by metasim** [comment]
Does this PR remove the ability to read directly from Tungsten memory, in cases where one wants to avoid the serialization/deserialization costs? IOW, in an `Expression` over a UDT, can one still pull from select subfields without deserializing the whole structure into JVM memory? 

This is an important optimization for us in RasterFrames:

https://github.com/locationtech/rasterframes/blob/b731865d9d170fedbf0ffbe1e257694f4030db85/core/src/main/scala/org/apache/spark/sql/rf/InternalRowTile.scala#L43-L53

Just want to look ahead to what we might need to change down the road.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UserDefinedType.scala [hunk 2]
```diff
@@ -50,13 +54,46 @@ abstract class UserDefinedType[UserType >: Null] extends DataType with Serializa
   /** Serialized Python UDT class, if exists. */
   def serializedPyClass: String = null
 
-  /**
-   * Convert the user type to a SQL datum
-   */
-  def serialize(obj: UserType): Any
+  /** The RowEncoder used to serialize/deserialize extenal row to internal row format. */
+  private lazy val rowEncoder = UserDefinedType.getRowEncoder(sqlType)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UserDefinedType.scala]: Due to recently added constraint that SQLConf can't be accessed during task execution, this causes the test failure. I'm wondering if we have other options here...

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by metasim** [comment]
Very excited to have this implemented!

**Comment by wetneb** [comment]
Could we please reopen this @viirya? It would be great to have this in 3.0!

**Comment by viirya** [comment]
> Could we please reopen this @viirya? It would be great to have this in 3.0!

I can reopen this. But this is stuck for a while so I'm not sure if this can be in...

**Comment by wetneb** [comment]
Thanks!

Is there a particular reason to hold this back @rxin @dongjoon-hyun? This PR was open more than 3 years ago now and there does not seem to be any reservations about it so far. When can we hope this to be released?

**Comment by dongjoon-hyun** [comment]
Hi, @viirya . When you reopen the PR, you need to remove `Stale` tag to prevent the bot.

**Comment by viirya** [comment]
@dongjoon-hyun Got it, thanks for the hint.

**Comment by wetneb** [comment]
Could someone reopen this and remove the "Stale" tag then?

**Comment by viirya** [comment]
I reopened it. But I'm pretty sure this cannot be in 3.0...

**Comment by wetneb** [comment]
If not in 3.0, let's hope that User Defined Types will be… definable by users, at some point!

**Comment by metasim** [comment]
The [RasterFrames](https://rasterframes.io/) project has had a [UDT](https://github.com/locationtech/rasterframes/blob/develop/core/src/main/scala/org/apache/spark/sql/rf/TileUDT.scala) at its core for around 3 years. Aside from having to be declared under the `org.apache.spark.sql` package as a hack around `UserDefinedType` being package private, it has worked quite well. Is there some context or forum where sharing our experience might help pave the way for this to become an official API?

**Comment by viirya** [comment]
Let's see how the community react to UDT now as I'm just returning back due to a new JIRA comment.

Seems there are a few code conflicts. I will take some time to solve it.

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by viirya** [comment]
I guess most people want is to just open UDT API. Close this in favor of the simple opening PR.
